sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala

```diff
@@ -17,18 +17,270 @@
+  }
+
+  def remove(): UnsafeRow = {
+    synchronized {
+      if (out != null) {
+        out.flush()
+        out.close()
+        out = null
+        fout.close()
+        fout = null
+
+        fin = new FileInputStream(path)
+        in = new DataInputStream(new BufferedInputStream(fin))
+      }
+    }
+
+    if (length > 0) {
+      val size = in.readInt()
+      assert(4 + size <= length, s"require ${4 + size} bytes for next row")
+      val bytes = new Array[Byte](size)
+      in.readFully(bytes)
+      length -= 4 + size
+      resultRow.pointTo(bytes, size)
+      resultRow
+    } else {
+      null
+    }
+  }
+
+  def close(): Unit = {
+    synchronized {
+      if (fout != null) {
+        fout.close()
+        fout = null
+      }
+      if (fin != null) {
+        fin.close()
+        fin = null
+      }
+    }
+    val file = new File(path)
+    if (file.exists()) {
+      file.delete()
+    }
+  }
+}
+
+/**
+ * A RowQueue that has a list of RowQueues, which could be in memory or disk.
+ *
+ * HybridRowQueue could be safely appended in one thread, and pulled in another thread in the same
+ * time.
+ */
+private[python] case class HybridRowQueue(
+    memManager: TaskMemoryManager,
+    dir: File, fields: Int)
+  extends MemoryConsumer(memManager) with RowQueue {
+
+  // Each buffer should have at least one row
+  private val queues = new java.util.LinkedList[RowQueue]()
+
+  private var writing: RowQueue = _
+  private var reading: RowQueue = _
+
+  private[python] def numQueues(): Int = queues.size()
+
+  def spill(size: Long, trigger: MemoryConsumer): Long = {
+    if (trigger == this) {
+      // When it's triggered by itself, it should write upcoming rows into disk instead of copying
+      // the rows already in the queue.
+      return 0L
+    }
+    var released = 0L
+    synchronized {
+      // poll out all the buffers and add them back in the same order to make sure that the rows
+      // are in correct order.
+      val n = queues.size()
+      var i = 0
+      while (i < n) {
+        i += 1
+        val queue = queues.remove()
+        val newQueue = if (i < n && queue.isInstanceOf[InMemoryRowQueue]) {
+          val diskQueue = createDiskQueue()
+          var r = queue.remove()
+          while (r != null) {
+            diskQueue.add(r)
+            r = queue.remove()
+          }
+          released += queue.asInstanceOf[InMemoryRowQueue].page.size()
+          queue.close()
+          diskQueue
+        } else {
+          queue
+        }
+        queues.add(newQueue)
+      }
+    }
+    released
+  }
+
+  private def createDiskQueue(): RowQueue = {
+    DiskRowQueue(File.createTempFile("buffer", "", dir).getAbsolutePath, fields)
+  }
+
+  private def createNewQueue(required: Long): RowQueue = {
+    val buffer = try {
+      val page = allocatePage(required)
+      new InMemoryRowQueue(page, fields) {
+        override def close(): Unit = {
+          freePage(page)
+        }
+      }
+    } catch {
```

Comment:
Could we narrow the scope of this `OOM`-catching to only wrap the `allocatePage` call? This would require us to declare `page` before `buffer` but that's fine.